compiler/rustc_const_eval/src/interpret/intern.rs

```diff
@@ -158,30 +140,79 @@ pub fn intern_const_alloc_recursive<
                 // Inner allocations are never mutable. They can only arise via the "tail
                 // expression" / "outer scope" rule, and we treat them consistently with `const`.
                 Mutability::Not,
-                true,
             )
         }
         InternKind::Static(Mutability::Mut) => {
             // Just make everything mutable. We accept code like
             // `static mut X = &mut [42]`, so even inner allocations need to be mutable.
-            (Mutability::Mut, Mutability::Mut, true)
+            (Mutability::Mut, Mutability::Mut)
         }
     };
+    let base_alloc_id = ret.ptr().provenance.unwrap().alloc_id();
+    let mut todo: Vec<_> = {
+        let base_alloc = &mut ecx.memory.alloc_map.get_mut(&base_alloc_id).unwrap().1;
+        base_alloc.mutability = base_mutability;
+        base_alloc.provenance().ptrs().iter().copied().collect()
+    };
+    let mut seen = FxHashSet::default();
+    seen.insert(base_alloc_id);
+    while let Some((_, prov)) = todo.pop() {
+        if !seen.insert(prov.alloc_id()) {
+            // Already processed
+            continue;
+        }
+        let Some((_, alloc)) = &mut ecx.memory.alloc_map.get_mut(&prov.alloc_id()) else {
+            continue;
+        };
+        // We always intern with `inner_mutability`, and furthermore we ensured above that if
+        // that is "immutable", then there are *no* mutable pointers anywhere in the newly
+        // interned memory -- justifying that we can indeed intern immutably. However this also
```

Comment:
Hm, we're not really ensuring this any more, are we?

Reply:
That check is now done by `intern_const_alloc_recursive`. And that function has to re-compute `inner_mutability` just for that purpose.

Maybe `patch_mutability_of_allocs` should still do the check, but just return whether there was a problem or not, then we run validation, and then if validation didn't error and there was a problem during `patch_mutability_of_allocs`, *then* we show the other error? Thinking about it, if we do that we can probably still do interning before validation... :see_no_evil:

Reply:
Yea, there were more follow-up cleanups to do, but I didn't want to change too much in one go

Reply:
If we follow the strategy of "do interning first, record if there was an error, but delay reporting that until after validation" -- that requires way fewer changes than this PR, doesn't it?

Reply:
Oh I remember now: we can make interning completely infallible by moving the `CONST_EVAL_MUTABLE_PTR_IN_FINAL_VALUE` into validation

Reply:
Ok, I implemented this in https://github.com/rust-lang/rust/pull/122684

I think we should still go with this PR, as it allows making individual things simpler, even if it means validation needs to handle allocations that haven't been interned yet. Most of the time that should be resolveable by using general interpreter methods instead of using dedicated `global_alloc` code paths

Reply:
I changed that in this PR. After a place has been validated, we also do some checks on all relocations of that place

Reply:
Hm, to me that doesn't fit with the intention of validation as a type-driven traversal... I'll have to think about this.